h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -0,0 +1,193 @@
+package water.persist;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.storage.*;
+import water.H2O;
+import water.Key;
+import water.MemoryManager;
+import water.Value;
+import water.api.FSIOException;
+import water.fvec.FileVec;
+import water.fvec.GcsFileVec;
+import water.util.Log;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.*;
+
+/**
+ * Persistence backend for GCS
+ */
+public final class PersistGcs extends Persist {
+
+  private static final Storage storage = StorageOptions.getDefaultInstance().getService();
+
+  @Override
+  public byte[] load(final Value v) throws IOException {
+    final BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    
+    final byte[] contentBytes = MemoryManager.malloc1(v._max);
+    final ByteBuffer wrappingBuffer = ByteBuffer.wrap(contentBytes);
+    final Key k = v._key;
+    long offset = 0;
+    // Skip offset based on chunk number
+    if(k._kb[0] == Key.CHK) {
+      offset = FileVec.chunkOffset(k); // The offset
+    }
+    
+    final ReadChannel reader = storage.reader(blobId);
+    reader.seek(offset);
```

Comment:
Does the seek actually skip the bytes? Or does it transfer them over the wire? (We had this issue with S3 before...)

Reply:
In our case, the [BlobReadChannel](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobReadChannel.java) is used. In case of BlobReadChannel, bytes are simply skipped.

```java
  @Override
  public void seek(long position) throws IOException {
    validateOpen();
    this.position = position;
    buffer = null;
    bufferPos = 0;
    endOfStream = false;
}
```

Reply:
Perfect! Thanks for confirming